### PR TITLE
Set up base share extension (EXP-79)

### PR DESCRIPTION
**Pull request by trevor-coleman**
# Set up base share extension (EXP-79)

This PR implements a minimal share extension for iOS that immediately sends the file without opening a react native view in the sharing app following the [iOS instructions](https://github.com/Expensify/react-native-share-menu/blob/master/IOS_INSTRUCTIONS.md) from `react-native-share-menu` with some changes outlined below.

It also implements a temporary hook (`useReactNativeShareMenuData()`) for testing that logs the received data to the console for testing purposes.

## Temporary Hook (`useReactNativeShareMenuData()`)

This should be pretty self explanatory, it receives the data either logs or throws warnings to indicate the state of the data.

The hook is the first thing called in the root component:

```js
// ~/src/App.js
//...

import useReactNativeShareMenuData from './hooks/useReactNativeShareMenuData';

//...

function App() {
    useReactNativeShareMenuData();     //  <-- 
    useDefaultDragAndDrop();
    OnyxUpdateManager();
    return (   
          //...
    )
    
```

This could become a permanent hook, if it returned the shared data, or we could use Ryan's `Provider` architecture, or some hybrid, but what the right solution is will depend on where in the app the share screen ends up living. Since the shared data is ephemeral (only used once, and only in one place) a provider may be unnecessary.

## Removing the content in the storyboard

Because we're not rendering the view, we removed most of the content in `ios/NewExpensifyShare/Base.lproj/MainInterface.storyboard`

I just followed the changes from @ryanlntn's branch (#63) on this one.

## Use our own ShareViewController rather than the one from RNSM

The instructions for RNSM say to reference the `ShareViewController` in `node_modules`, but the PR uses our own copy in the `NewExpensifyShare` folder.

This allows two things that enable the "no UI" flow:

1. It lets us have `ShareViewController` inherit from `UIViewController`
2. We can move the `self.didSelectPost()` call into the `viewDidLoad` method, so it fires immediately without user interaction.

## RNSM Patch: App Group Naming

The Expensify app uses a couple of bases for its bundle IDs: `com.expensify.chat` and `com.chat.expensify.chat`. The different build configurations (DebugDevelopment, DebugProduction, etc) are distinguished by post-fixes on these themes such as `com.chat.expensift.chat.dev` or `com.expensify.chat.adhoc`.

RNSM assumes that all apps will follow the reverse-domain naming scheme that apple recommends (which Expensify does not), so it assumes that the group name for the app will be `group.{bundleID}`. 

That means it assumes the group for `DebugDevelopment` will be `group.com.expensify.chat.dev` but the way it is currently set up we are using one AppGroup ID for all of the different configs `group.com.chat.expensify.chat`

### Current temporary solution

We've currently solved the app group issue in one of two ways at various times:

1. hard-coded the group ID into the swift code, or
2. referenced a key in the app's `info.plist` that has the group ID
     - The key is called `ReactNativeShareMenuAppGroup`
     ```
     <key>ReactNativeShareMenuAppGroup</key>
	<string>group.com.chat.expensify.chat</string>
     ``` 
     - there is a constant in the `RNShareMenu` pod that tells the extension where to find it
    ```swift
    
    public let HOST_APP_GROUP_INFO_PLIST_KEY:String = "ReactNativeShareMenuAppGroup"
    ```

#### Patching RNSM

A [future ticket](https://www.notion.so/Fix-how-RNSM-handles-app-groups-36546faf3dc648d2aa0b5278330eac57?pvs=4) already exists to determine a better and more stable way of finding and referencing app groups. We should probably wait until we've decided how the library is going to work before wewe spend too much effort cleaning this up.

We should also look at how to have the app group names properly reflect the bundle IDs of the various environments, and how to do that in ways that minimize the duplication of files -- one key to that is the `CODE_SIGN_ENTITLEMENTS` key in the pbxproj.

#### CODE_SIGN_ENTITLEMENTS
Right now there isa single entitlements file called `Chat.entitlements`, but if you manually edit the app group for one of the environments, then you'll get a new file called `NewExpensifyDebugDevelopment.entitlements` or whatever. 

I was looking into how this is set up, and the `CODE_SIGN_ENTITLEMENTS` key apparently points to the entitlements file.

We could probably set up a second file -- something like `Debug.entitlements`. Then that could be for the `com.expensify` environments, while the `Chat.entitlements` file could handle the `com.expensify` domains. 

I don't know if it's possible to set the name of that file through the XCode UI but you can easily just edit the `CODE_SIGN_ENTITLEMENTS` key in the `pbxproj`.

I'm also not sure if you need to have the complete file duiplicated, or if the file works as an override with some fallback, but something to look into.

### CODE REVIEW

**Review by ryanlntn — COMMENTED** [review]
> This could become a permanent hook, if it returned the shared data, or we could use Ryan's Provider architecture, or some hybrid, but what the right solution is will depend on where in the app the share screen ends up living. Since the shared data is ephemeral (only used once, and only in one place) a provider may be unnecessary.

@trevor-coleman That makes sense to me. The reason I went with react context here was that we can only get share data once from RNSM and we had multiple screens we needed to access it in and Expensify's navigation doesn't make it particularly easy to pass as route params.

**Comment by lindboe** [comment]
Big comment covering the writeup and discussion so far:

> This PR implements a minimal share extension for iOS that immediately sends the file without opening a react native view in the sharing app following the [iOS instructions](https://github.com/Expensify/react-native-share-menu/blob/master/IOS_INSTRUCTIONS.md) from react-native-share-menu with some changes outlined below.

I think it's worth calling out that we're also patching the default RNSM ViewController and why (UIViewController allows us to avoid showing the default OS prompt, and we fire the `post` action automatically), and that we want to follow up with a PR to RNSM to support this "No UI" flow officially.

> The Expensify app uses a couple of bundle IDs: com.expensify.chat and com.chat.expensify.chat

Technically this is not quite accurate; the bundle ID is `com.expensify.chat.dev`, etc. `com.expensify.chat` is not a bundle ID anywhere that I know of, but rather a base pattern the new configuration uses.

> We could probably set up a second file -- something like Debug.entitlements. Then that could be for the com.expensify environments, while the Chat.entitlements file could handle the com.expensify domains.

I'm not sure of the purpose this suggestion serves?

> The reason I went with react context here was that we can only get share data once from RNSM and we had multiple screens we needed to access it in and Expensify's navigation doesn't make it particularly easy to pass as route params.

Just for context, the Expensify team seems fine with updating the navigation utils to make it possible to pass route params. Unless @ryanlntn is referring to a different problem?

**Comment by ryanlntn** [comment]
> When you say it's only available once--does it get erased immediately after it is read?

There's just not a way to get the share data after the initial read. `getInitialShare` only returns data on the first call and `addNewShareListener` only fires once so if you want to access the data via multiple instances of a hook you need some way to persist that data or else pass it along.

**Comment by lindboe** [comment]
> There's just not a way to get the share data after the initial read.

@ryanlntn Do you know, does this mean RNSM has the same issue with UserDefaults as with copied files, that the data never gets deleted? Or is it overwritten on future shares? (If you don't know, no worries, I'll go look)

**Comment by ryanlntn** [comment]
> @ryanlntn Do you know, does this mean RNSM has the same issue with UserDefaults as with copied files, that the data never gets deleted? Or is it overwritten on future shares? (If you don't know, no worries, I'll go look)

It just overwrites the same key in UserDefaults every time so there isn't an issue like with copying files. This is more just due not having a way to directly access that last shared data. Actually `ShareMenuReactView` does have a method to access this data imperatively so maybe that could be added to RNSM itself as another workaround.

**Comment by trevor-coleman** [comment]
> > We could probably set up a second file -- something like Debug.entitlements. Then that could be for the com.expensify environments, while the Chat.entitlements file could handle the com.expensify domains.
> 
> I'm not sure of the purpose this suggestion serves?

We eventually should use `group.com.chat.expensify.chat` for bundleIDs that start with `com.chat.expensify.chat` and `group.com.expensify.chat` for bundleIDs that start with `com.expensify.chat`

The base way to do this would be to have a different file for each of the environment. So we'd have `DebugDevelopment.entitlements` and `DebugProduction.entitlements` etc etc etc

But you can instead point to a file, so we can group the similar cases into a single file. Right now they all point to `Chat.entitlements` but we can make the ones that need to be different point to a second file, which I would think would be called `Debug` but  could be called whatever.

**Comment by trevor-coleman** [comment]
@lindboe updated the writeup to include a description of the `ShareViewController` changes

**Comment by trevor-coleman** [comment]
@linboe -- if you can approve this I'll get it merged and can keep pushing this forward.

File: ios/NewExpensifyShare/ShareViewController.swift [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+import Foundation
+import os
+import MobileCoreServices
+import UIKit
+import Social
+import RNShareMenu
+
+
+
+class ShareViewController: UIViewController {
```
Comment (lindboe) [inline ios/NewExpensifyShare/ShareViewController.swift:14]: Just linking back to the relevant comment on this, but it's not a blocker: https://github.com/infinitered/ExpensifyApp/pull/65#discussion_r1385742814

**Review by lindboe — APPROVED** [review]
I do still want @cdanwards input on how clear this is, but we don't have to block merge on that, we can edit/follow up later.

**Pull request merged**